python/runtime/pai/entry/tf.py

```diff
@@ -0,0 +1,74 @@
+# Copyright 2020 The SQLFlow Authors. All rights reserved.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import pickle
+import types
+from inspect import getargspec
+
+from runtime.pai import train
+from runtime.pai.pai_distributed import define_tf_flags, set_oss_environs
+from runtime.tensorflow import explain, is_tf_estimator, predict
+from runtime.tensorflow.diag import SQLFlowDiagnostic
+
+
+def call_fun(func, params):
```

Comment:
`call_func`?

By the way, I don't like this function, it's hard to read. Can we have generic runtime API like `runtime.pai.train(datasource, select, feature_columns, ...)` and call the API by go side code generation?

Reply:
This is an entry point function on PAI, programmer do not call this function directly. They call [submit_pytf_train](https://github.com/sql-machine-learning/sqlflow/blob/30c90f6e47135111c37eb28ca89c569b57fc202f/python/runtime/pai/submitter.py#L324) after feature derivation and other things. `submit_pytf_train` wrap params and resources. On PAI the entry function unzip the resource and call specific train/predict/explain by this `call_fun`. It is a dispatcher, all param checks have already done in function like submit_pytf_train. If we don't use this way, we just expand every params, and that is equivalent to this.